hail/src/main/scala/is/hail/utils/Interval.scala

```diff
@@ -198,182 +197,20 @@ object Interval {
       }
     }
   }
-}
-
-case class IntervalTree[U: ClassTag](root: IntervalTreeNode[U]) extends
-  Traversable[(Interval, U)] with Serializable {
-  override def size: Int = if (root == null) 0 else root.size
-
-  def isEmpty(pord: ExtendedOrdering): Boolean = root == null
-
-  def contains(pord: ExtendedOrdering, position: Any): Boolean = root != null && root.contains(pord, position)
 
-  def overlaps(pord: ExtendedOrdering, interval: Interval): Boolean = root != null && root.overlaps(pord, interval)
-
-  def isDisjointFrom(pord: ExtendedOrdering, interval: Interval): Boolean = root == null || root.isDisjointFrom(pord, interval)
-
-  def queryIntervals(pord: ExtendedOrdering, position: Any): Array[Interval] = {
-    val b = Array.newBuilder[Interval]
-    if (root != null)
-      root.query(pord, b, position)
-    b.result()
-  }
-
-  def querySingleValue(pord: ExtendedOrdering, position: Any): U = {
-    root.querySingleValue(pord, position)
-  }
-
-  def queryValues(pord: ExtendedOrdering, position: Any): Array[U] = {
-    val b = Array.newBuilder[U]
-    if (root != null)
-      root.queryValues(pord, b, position)
-    b.result()
-  }
-
-  def queryOverlappingValues(pord: ExtendedOrdering, interval: Interval): Array[U] = {
-    val b = Array.newBuilder[U]
-    if (root != null)
-      root.queryOverlappingValues(pord, b, interval)
-    b.result()
-  }
-
-  def foreach[V](f: ((Interval, U)) => V) {
-    if (root != null)
-      root.foreach(f)
-  }
-}
-
-object IntervalTree {
-  def fromOption[U: ClassTag](root: Option[IntervalTreeNode[U]]): IntervalTree[U] = new IntervalTree(root.orNull)
-
-  def annotationTree[U: ClassTag](pord: ExtendedOrdering, values: Array[(Interval, U)]): IntervalTree[U] = {
-    val iord = Interval.ordering(pord, startPrimary = true)
-    val sorted = values.sortBy(_._1)(iord.toOrdering.asInstanceOf[Ordering[Interval]])
-    IntervalTree.fromOption[U](fromSorted(pord, sorted, 0, sorted.length))
-  }
-
-  def apply(pord: ExtendedOrdering, intervals: Array[Interval]): IntervalTree[Unit] = {
-    val iord = Interval.ordering(pord, startPrimary = true)
-    val sorted = if (intervals.nonEmpty) {
-      val unpruned = intervals.sorted(iord.toOrdering.asInstanceOf[Ordering[Interval]])
-      var i = 0
-      val ab = new ArrayBuilder[Interval](intervals.length)
-      var tmp = unpruned(i)
-      while (i < unpruned.length) {
-        tmp.merge(pord, unpruned(i)) match {
-          case Some(interval) =>
-            tmp = interval
-          case None =>
-            ab += tmp
-            tmp = unpruned(i)
-        }
-        i += 1
+  /** Comparison of 'x' and 'y', each of which is either an interval whose
+    * point type is comparible with 'pord', or an arbitrary annotation
+    * comparible with 'pord'. Returns true if 'x' is completely below 'y'.
+    */
+  def quasiordering(pord: ExtendedOrdering)(x: Any, y: Any): Boolean = {
```

Comment:
this makes me uncomfortable -- there's no reason that an interval's point type cannot also be interval, which makes this implementation ambiguous.

Reply:
The messy case is when we want to do binary search in an array of intervals (a partitioner) with a non-interval key, like when making a Spark partitioner, to find the partition containing a key.

What I want to do, and could do in C++ (say), is just search with the interval [k, k], but we can't do that without allocating. The only other way I see is to pass the binary search method two comparisons, so if we're searching in an `Array[T]` using a search key of type `K`, we'd need `comp(T, K)` and `comp(K, T)`. That's not ideal, but it's not so bad, I'll probably just do that.

Reply:
I got rid of this by making the binary search methods allow passing mixed-type comparisons. I agree it's better this way, though it comes at the cost of making the binary search implementation harder to understand.